### PR TITLE
add native histogram exemplar support

### LINKED ISSUE #1126: histograms: Support exemplars in pure Native Histograms
This depends on https://github.com/prometheus/client_model/issues/61 .

Exemplars can be added to the buckets of conventional histograms but not to native histograms. So far, the way of sending exemplars is by also rendering a conventional histogram. With the issue above fixed, we have a way of sending exemplars independently of conventional buckets. However, without conventional buckets, we don't have a heuristics which exemplars to pick. Native histograms can have so many buckets that adding an examplar to each is infeasible.

This is P2 as explained in the [milestone's description](https://github.com/prometheus/client_golang/milestone/6).

### PR DESCRIPTION
**Pull request by fatsheep9146**
fix #1126

### CODE REVIEW

File: prometheus/histogram.go
Comment (ArthurSens) [inline prometheus/histogram.go:834]:
```suggestion
		his.Exemplars = make([]*dto.Exemplar, len(h.nativeExemplars.exemplars))
```
We could pre-allocate the array size here :)

File: prometheus/histogram_test.go
Comment (ArthurSens) [inline prometheus/histogram_test.go:1345]: the histogram object has the `now()` function, I imagine we can prevent using sleep here if we just override the histogram now function to something like `h.now = func() time.Time { now.Add(time.Second * 10) }`

**Review by ArthurSens — COMMENTED** [review]
not a native histogram expert here, just some small comments :)

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1613]: These two need reasonable defaults in case they are left at their zero value (something like 5m for the TTL and 10 for the count?). And the defaults need to be documented in the doc comment for the options, too.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1620]: So the problem here is that this isn't concurrency safe. Note how the exemplars for classic exemplars are done (with a `[]atomic.Value`).

In the worst case, you could put everything behind a mutex, but that's quite slow. One might argue that `ObserveWithExemplar` shouldn't be called with high frequency anyway, and the mutex is OK for it (but that needs to be documented, should we go down that path).

I'm not sure if there is a feasible way to do something faster than the full mutex protection. For classic buckets, it's easy to know which exemplar to replace, but with the algorithm here, it's a much more complex decision.

Maybe let's do the mutex approach first and then try something more sophisticated if need arises.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1602]: While the algorithm seems to do the right thing, and your implementation might be the most efficient one in terms of big-O runtime, I do believe that it has a lot of baseline overhead, with two linked list and a map and many many pointers all over the place. Pointers have to be traversed, and they also take a lot of memory, which matters because an instrumented program might create many histograms.

Given that only a moderate number of exemplars make sense fundamentally (because that's the whole idea of exemplars to just present a few example observations), I'm pretty sure that an approach with a single slice will be more efficient both in terms of actual runtime and memory footprint.

So maybe have a single slice of size `nativeHistogramMaxExemplarCount` and sort it by the observed value. You can iterate through all ~10 elements once, searching for the oldest element and for the two closest ones at the same time. That should be super fast in practice. Insertion into the slice requires moving all the elements above by one, which is something computers can do super fast. (See https://go.dev/wiki/SliceTricks for trick to do it fast.)

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thanks for doing this. I think this implements the correct algorithm, but there is the huge problem of concurrency. See comments. A also think a more basic approach with a single slice rather than two linked list and a map will be faster in practice and consume less memory. Also see comments.

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go:1602]: > While the algorithm seems to do the right thing, and your implementation might be the most efficient one in terms of big-O runtime, I do believe that it has a lot of baseline overhead, with two linked list and a map and many many pointers all over the place. Pointers have to be traversed, and they also take a lot of memory, which matters because an instrumented program might create many histograms.
> 
> Given that only a moderate number of exemplars make sense fundamentally (because that's the whole idea of exemplars to just present a few example observations), I'm pretty sure that an approach with a single slice will be more efficient both in terms of actual runtime and memory footprint.
> 
> So maybe have a single slice of size `nativeHistogramMaxExemplarCount` and sort it by the observed value. You can iterate through all ~10 elements once, searching for the oldest element and for the two closest ones at the same time. That should be super fast in practice. Insertion into the slice requires moving all the elements above by one, which is something computers can do super fast. (See https://go.dev/wiki/SliceTricks for trick to do it fast.)

I understand your point, I think this is better since the number of exemplars may be only around 10, I will change the code like you said. @beorn7

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go:741]: I'm a little confused here, do you mean using pointer will create more overhead than not using pointer? @beorn7

Comment (beorn7) [reply prometheus/histogram.go:741]: Yes. It's an additional 8 bytes in the struct, and it has to be dereferenced for each access.

The only upside of a pointer is that you don't need to copy the whole struct the pointer points to if you hand things around. But computers are super fast when it comes to memcopying, and also we are not really handing around the `nativeExemplars` struct a lot here.

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go]: @beorn7 
I already changed the implementation from linked list into slice, but I did not keep the slice sorted as you suggested, here are my thoughts about this  

1. We should always iterate all exemplars to get the exemplar with oldest timestamp, so the time efficient should always be O(n)
2. If we do not keep the slice sorted, we can just change the exemplar in-place with new exemplar, so there is not need to adjust the slice with like  https://go.dev/wiki/SliceTricks does, which makes the code easier.

WDYT?

BTY, I will fix your other reviews when we can agree with this implementation.

Comment (beorn7) [reply prometheus/histogram.go]: See my comment below. I think we don't want to find the exemplar closest to the new exemplar, but the pair of exemplars closest to each other (including the new exemplar) and then delete the older of the two in the couple.

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go:1631]: > This code selects the existing exemplar that is closest to the new exemplar. However, my idea was to find the two exemplars closest to each other (including the new one, but also all existing one). Finding that pair is O(n²), unless you keep the slice sorted, in which case it is O(n).

Got you, I will change the code.

Comment (fatsheep9146) [reply prometheus/histogram.go]: The code is refracted, please help review it again when you have time. @beorn7

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go:1627]: added a `sync.Mutex` to `nativeExemplars` struct

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go:1621]: add default value and comments in doc.

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go:487]: Done

**Comment by beorn7** [comment]
I'll try to get to this ASAP, but currently, I'm sick.

**Comment by fatsheep9146** [comment]
> I'll try to get to this ASAP, but currently, I'm sick.

Thanks! Please take good care of yourself and rest well, there is no rush

**Review by kakkoyun — COMMENTED** [review]
Nice to have this. Thanks for the contribution 🎉

**Review by beorn7 — CHANGES_REQUESTED** [review]
Thanks a lot. I finally have done a detailed review. Things are a bit tricky, but I hope I have explained it well enough.

In different news, the doc comment for the `ExemplarObserver` interface (in file observer.go) needs an update, explaining broadly the behavior for native histograms, which should also include a warning that `ObserveWithExemplar` isn't lock-free if called on a native histogram with configured exemplars so that it should not be called too frequently.

**Comment by fatsheep9146** [comment]
@beorn7 The pr is ready for review again.

**Comment by beorn7** [comment]
Thank you very much. I'm on vacation this week, but this is high on my list once I'm back in action.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:754]: Nit: Maybe move this up to just after `resetScheduled` as `now` and `afterFunc` are just for testing purposes, while `nativeExemplars` is important for regular use.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1135]: The doc comment for this method (lines 1121-1122) needs an update to mention that nativeExemplars are handled, too.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1615]: Nit: I see no reason to use a blank line between exemplars and ttl. With only three members, one might even argue that the mutex shouldn't be separated.

I have a minor preference to list a mutex that protects the whole struct first, but I'm not sure if that's really a convention we generally follow in the code base.

In summary:

```suggestion
	sync.Mutex

	ttl       time.Duration
	exemplars []*dto.Exemplar
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1649]: maxCount → cap(n.exemplars)

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1669]: I think those comments are easier to do inline. Suggestion:

```suggestion
		ot    = time.Now() // Oldest timestamp seen.
		otIdx = -1         // Index of the exemplar with the oldest timestamp.
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1677]: In the same spirit:

```suggestion
		md    = -1.0 // Logarithm of the delta of the closest pair of exemplars.
		mdIdx = -1   // Index of the older exemplar within the closest pair.
		cLog float64 // Logarithm of the current exemplar.
		pLog float64 // Logarithm of the previous exemplar.
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1681]: Language nit:
```suggestion
		// Find the exemplar with the oldest timestamp.
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1664]:
```suggestion
		rIdx int // The index where to remove the old exemplar.
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1687]: Language nit:
```suggestion
		// Find the index at which to insert new the exemplar.
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1692]:
```suggestion
		// Find the two closest exemplars and pick the one the with older timestamp.
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1711]:
```suggestion
	// then the exemplar should be inserted at the end.
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1721]:
```suggestion
		// In the previous for loop, when calculating the closest pair of exemplars,
		// we did not take into account the newly inserted exemplar.
		// So we need to calculate with the newly inserted exemplar again.
```

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1745]:
```suggestion
	// Adjust the slice according to rIdx and nIdx.
```

**Review by beorn7 — CHANGES_REQUESTED** [review]
Very cool.

I think all that remains is wordsmithing and some formatting tweaks. See all my pedantic comments. :)

**Comment by fatsheep9146** [comment]
@beorn7 
The commit history seems too long,  I can squash to one commit if you think there are no more changes needed.

**Comment by beorn7** [comment]
No worries, I can just squash during merge.

Or, if you like a more detailed commit structure with more than one commit, but not the 20 we have now, you could rebase on your own.

**Comment by beorn7** [comment]
One thing fell through the cracks: The doc comment to the ObserveWithExemplar method in histogram.go should contain a warning that it  isn't lock-free for native histograms with configured exemplars and therefore should not be called in high-frequency settings.

**Comment by fatsheep9146** [comment]
@beorn7 I only kept two commits, one for first implementation, the other for reconstitution that you suggested in https://github.com/prometheus/client_golang/pull/1471#discussion_r1559711441

File: prometheus/histogram.go
Comment (bwplotka) [inline prometheus/histogram.go:482]:
```suggestion
	NativeHistogramMaxExemplars int
```

Comment (bwplotka) [reply prometheus/histogram.go:482]: Can we simplify? (also ditto name in commentary)

**Review by bwplotka — APPROVED** [review]
Happy to get this in, just one suggestion. Thanks!!! Sorry for lag on review.

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go]: Done

**Comment by bwplotka** [comment]
It feels @beorn7 comments are resolved, did you intent @beorn7 to approve it after you requested changes? https://github.com/prometheus/client_golang/pull/1471#pullrequestreview-2036231494

**Comment by fatsheep9146** [comment]
> It feels @beorn7 comments are resolved, did you intent @beorn7 to approve it after you requested changes?

Yes, all comments are resolved, but I think @beorn7 is waiting for other reviewers to raise objections?

> Otherwise, this looks good. I'll give @ArthurSens @bwplotka and @kakkoyun a few more days to raise objections.

File: prometheus/histogram.go
Comment (bwplotka) [inline prometheus/histogram.go:846]: This might be taken as a mistake, so let's have a quick commentary on why `cap` is used not `len` here, WDYT?

Comment (fatsheep9146) [reply prometheus/histogram.go:846]: We use `cap` here, because we first define that, if users use a negative `NativeHistogramMaxExemplars` value, then no exemplars will be cached. (You can see this in the comment of `NativeHistogramMaxExemplars`).

So when we call `makeNativeExemplars` to initialize a new `nativeExemplars`, if `NativeHistogramMaxExemplars` is negative, we just create a slice of exemplars with cap 0 to indicates that no exemplars will be cached.

So when `Write` is called, we use cap to check if user wants to cache exemplars. If they wants, then we will output the cached exemplars. 

But I think your idea is better, because when len is 0, although cap maybe greater than 0, we could still skip the following append operation since there are no cached exemplars at that time.

@bwplotka

**Comment by fatsheep9146** [comment]
> Well, but "Change Requested" status is still there. And that from @beorn7 .... one never should mess with the red mark from @beorn7 =DDDD
> 
> On serious note, I wanted to be lazy and trust @beorn7 to approve the most complex part of this PR, so `addExemplar` body properly, I only skimmed through that. Let me check that next week (I have to focus on other work now), if @beorn7 won't have time.
> 
> Is that ok? We can also merge and iterate... it's a new functionality.

I think it's better to do a more detailed review of this pr when you have time next week. I'm definitely ok with that. @bwplotka

File: prometheus/histogram_test.go
Comment (ArthurSens) [inline prometheus/histogram_test.go:1336]: This part repeats 3x in this test, maybe we can extract into a separate function just to reduce a few lines?

File: prometheus/histogram_test.go
Comment (ArthurSens) [inline prometheus/histogram_test.go:1317]: histograms have a field called `now`, their intended usage is exactly to mock time and avoid sleeps in tests.

You could do something like this when creating the histogram
```go
now := time.Now()
h.now = func time.Time { return now }
```

in `addFunc` you can change the returned value to:
```go
h.now = func time.Time { return now.Add(time.Second * 10) }
```

this would pretend that 10 seconds have passed, without needing to wait 10 seconds :)

**Review by ArthurSens — COMMENTED** [review]
I just a quick read through the tests and I already have to leave 😬 

I'll try to take another look at `addExamplar` tomorrow but I'm similar to bartek here, I trust Beorn's review and there's no need to wait for me if we want to merge it already

File: prometheus/histogram.go
Comment (beorn7) [reply prometheus/histogram.go:846]: If exemplars are configured, and we append the first exemplar, the `cap` is > 1, but the `len` is still 0. So it is correct and actually required to test for `cap` here.

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:765]: Slight clarification:

ObserveWithExemplar should not be called in a high-frequency setting for a native histogram with configured exemplars. For this case, the implementation isn't lock-free and might suffer from lock contention.

**Comment by beorn7** [comment]
I think @ArthurSens 's comments about the tests are valid. @fatsheep9146 I think you should address them.

Otherwise, I only have a slight clarification for the `ObserveWithExemplar` doc comment (to make clear that it is still lock-free if called for a native histogram without exemplars or for a classic histogram).

And then this is good to go. Thanks for all your efforts and patience. This was quite involved.

File: prometheus/histogram.go
Comment (ArthurSens) [reply prometheus/histogram.go:846]: From Bartek's comment I believe he agrees with the usage of `cap`, he just requested a comment on top so we can remember the reason in the future

Comment (fatsheep9146) [reply prometheus/histogram.go:846]: @ArthurSens I think you are right :)
I misunderstood @bwplotka ‘s comments.

File: prometheus/histogram_test.go
Comment (fatsheep9146) [reply prometheus/histogram_test.go:1336]: I convert the compare logic into one common function `compareNativeExemplarValues`

File: prometheus/histogram_test.go
Comment (fatsheep9146) [reply prometheus/histogram_test.go:1317]: Done

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go:765]: Done

File: prometheus/histogram.go
Comment (fatsheep9146) [reply prometheus/histogram.go:846]: I added a quick comment.

**Review by beorn7 — APPROVED** [review]

**Comment by beorn7** [comment]
Great. I'll merge on green.

**Pull request merged**
